features/graph/provider/application/src/main/java/org/opennms/netmgt/graph/provider/application/ApplicationGraphProvider.java

```diff
@@ -0,0 +1,123 @@
+import org.opennms.netmgt.graph.simple.SimpleGraph;
+import org.opennms.netmgt.graph.simple.SimpleVertex;
+import org.opennms.netmgt.model.OnmsApplication;
+import org.opennms.netmgt.model.OnmsMonitoredService;
+import org.opennms.netmgt.xml.event.Event;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class ApplicationGraphProvider implements GraphProvider, EventListener {
+
+    static final String TOPOLOGY_NAMESPACE = "application";
+
+    private static final Logger LOG = LoggerFactory.getLogger(ApplicationGraphProvider.class);
+
+    private Graph graph;
+
+    private ApplicationDao applicationDao;
+
+    private boolean initialized = false;
+
+    public ApplicationGraphProvider(ApplicationDao applicationDao) {
+        Objects.requireNonNull(applicationDao);
+        this.applicationDao = applicationDao;
+        LOG.debug("Creating a new {} with namespace {}", getClass().getSimpleName(), TOPOLOGY_NAMESPACE);
+    }
+
+    // TODO MVR We may need some kind of caching strategy implementation allowing each provider to deal with reloads individually if they so choose
+    @Override
+    public Graph<?, ?> loadGraph() {
+        // TODO MVR this is not thread safe
+        if (!initialized) {
+            graph = createGraph();
+            initialized = true;
+        }
+        return graph;
+    }
+
+    @Override
+    public GraphInfo<?> getGraphInfo() {
+        final DefaultGraphInfo graphInfo = new DefaultGraphInfo(TOPOLOGY_NAMESPACE, SimpleVertex.class);
+        graphInfo.setLabel("Application Graph");
+        graphInfo.setDescription("This Topology Provider displays the hierarchy of TODO."); // TODO Patrick
+        return graphInfo;
+    }
+
+    Graph<?, ?> createGraph() {
+        final SimpleGraph graph = new SimpleGraph(getGraphInfo());
+
+
+        for (OnmsApplication application : applicationDao.findAll()) {
+            ApplicationVertex applicationVertex = new ApplicationVertex(application);
+            graph.addVertex(applicationVertex);
+
+            for (OnmsMonitoredService eachMonitoredService : application.getMonitoredServices()) {
+                final ApplicationVertex serviceVertex = new ApplicationVertex(eachMonitoredService);
+                applicationVertex.addChildren(serviceVertex);
+                graph.addVertex(applicationVertex);
+
+                // connect with application
+                String id = String.format("connection:%s:%s", applicationVertex.getId(), serviceVertex.getId());
+                SimpleEdge edge = new SimpleEdge(applicationVertex, serviceVertex);
+                graph.addEdge(edge);
+            }
+        }
+        return graph;
+    }
+
+    @Override
+    public String getName() {
+        return getClass().getSimpleName();
+    }
+
+    @Override
+    public void onEvent(Event e) {
+        // BSM has been reloaded, force reload
+        // TODO: patrick find out if we need to listen to an event and for which one?
```

Comment:
to be discussed with @mvrueden

Reply:
Not necessary. It looks like you were copying the GraphProvider for BSM.
The idea behind the `GraphProvider`/`GraphContainerProvider` is, that it knows how to load a graph/container. On the other hand, the `GraphService` is invoking the loading methods accordingly. Also it should implement caching strategies. However each GraphProvider may need other caching strategies:

 - GraphML: Only load once, never reload
 - BSM: Only load when changed (thus the `onEvent(...)`
 - Application: Reload when cache is evicted (e.g. by default 5 Minutes)
 - Vmware: Maybe listen for updates and update adhoc (no reloading needed, only load once)
 - Linkd: See Vmware
 - and so on

However at this state caching is not implemented.
I decided to implement the "only reload on change" manually in BSM to add according TODOs to implement proper caching strategies at some point.

Reply:
ok, I added caching to ApplicationGraphProvider for now which we can remove again later when we have a generic caching in place

Reply:
Let's not add caching for each provider.
Everything goes into the dedicated feature branch and will only be merged when finished.
As caching strategies are a core concept it needs to be implemented before we can call this feature finished. There is no point in adding additional caching strategies for the providers, as that is dead code anyways.